SLIC MPI Collective Abort

bmhan12 opened the pull request:
This PR:

- Removes `abortIfEnabled()`/`flushStreams()` call from `logMessage()` (`logMessage() `is no longer collective)
  - For Error and Warning messages, `logMessage()` can now set an abort flag (`setAbortFlag()`). 
- `abortIfEnabled()` now does a collective call (`confirmAbortStreams()`) to check if any rank has set an abort flag before flushing and aborting.
  -  MPI_Allreduce is done in the MPI LogStreams (SynchronizedStream and LumberjackStream)
- Disables abort on error for certain examples that use MPI LogStreams.
  - Otherwise, MPI will hang program run when ranks branch off and call `SLIC_ASSERT` non-collectively

This PR addresses #26 

___

Updates:

- Removed collective call to check ranks for abort flags before flushing. The aborting rank now flushes locally then calls `MPI_Abort()`.
- `abortIfEnabled()` simplified to `abort()`, which just calls system abort or `MPI_Abort()`.
- Re-enabled abort on error for examples that were disabled before in this PR.

agcapps commented on src/axom/slic/interface/slic.cpp:
```diff
@@ -52,7 +52,7 @@ void ensureInitialized()
       message::Warning,
       "slic::initialize() must be called before any other calls to SLIC\n."
-      "The SLIC library called slic::initialize() for you and set up a minimal"
+      "The SLIC library called slic::initialize() for you and set up a minimal "
```
Thanks.

agcapps commented on src/axom/slic/interface/slic_macros.hpp:
```diff
@@ -70,6 +71,10 @@
       __oss << msg;                                                 \
       axom::slic::logErrorMessage(__oss.str(), __FILE__, __LINE__); \
     }                                                               \
+    if(axom::slic::isInitialized())                                 \
+    {                                                               \
+      axom::slic::abortIfEnabled(axom::slic::message::Error);       \
+    }                                                               \
```
Two related points:
- Shouldn't the call to `abortIfEnabled` only happen if `EXP` is true?  So should go inside the close-bracket at (new) line73?
- `slic::logErrorMessage()` eventually calls `ensureInitialized()` at slic.cpp (new) line 221, so we should not need to test on the result of `isInitialized()`.

bmhan12 replied:
- `abortIfEnabled` is how we collectively check that a rank has aborted. If we put  `abortIfEnabled()` inside the `EXP` block, then every rank must have `EXP` evaluate to the same value to call `abortIfEnabled`  or MPI will likely hang the application. The name `"abortIfEnabled"` is potentially not so clear, and does not stress that it must to be called collectively... Open to suggestions for names to all of the added "*abort*"  functions.
- Without the  `isInitialized()` block,` ensureInitialized() `will always be called from  `abortIfEnabled()`, even when no message has been logged by SLIC. This leads to `ensureInitialized()` warning messages for almost  every unit test that eventually hits a `SLIC_ASSERT`.

agcapps replied:
I think the naming of `abortIfEnabled` was part of what confused me, though it should not have.  Maybe consider one of 
- abortIfSignaled (hmmm, might connote operating system signals)
- abortIfFlagged
- abortIfRequested (like the stop request pull cord on a bus?)
- abortIfSomeoneRequested (eh, that's getting silly)

Thank you for explaining both points.

white238 replied:
`abortIfAbortOnWarningOrErrorIsEnabled()`?

bmhan12 replied:
Hmmm....  thinking on it more, `abortIfEnabled(level)` fits well with the `enableAbortOnXXX` and `disableAbortOnXXX` functions. I like the verbs `enable/disable` for these functions (I can't think of a better replacement), and it makes sense for `abortIfXXX` to use the same verbiage. Kind of convinced myself to keep it the same.

kennyweiss replied:
My thinking is more inline with @agcapps.

How about something like this:

```cpp
  do
  {
    if(EXP)
    {
      std::ostringstream __oss;
      __oss << msg;
      // Note: the following internally calls Logger::setAbortFlag() when necessary
      // using Logger's m_abortOnError / m_abortOnWarning flags
      axom::slic::logErrorMessage(__oss.str(), __FILE__, __LINE__);       
    }
    // Note: Abort will only be raised locally/globally when m_abortOnError and/or m_abortOnWarning
    // So, the following will only need to be collective when m_abortOnError
    // and the underlying LogStream is MPI-based
    if(axom::slic::checkAbortRaised(message::Level))
    { 
      axom::slic::flushStreams();  // note: all ranks must already participate for MPI-based LogStreams
      axom::slic::abort();   // note: no message::Level
    }
  } while(axom::slic::detail::false_value)
```

kennyweiss replied:
@bmhan12 wrote:
> Without the isInitialized() block,ensureInitialized()will always be called from abortIfEnabled(), even when no message has been logged by SLIC. This leads to ensureInitialized() warning messages for almost every unit test that eventually hits a SLIC_ASSERT, with nothing actually logged.

We should not have any unit tests that use slic without initializing it first.
I'd argue that this is a feature, and we should fix any test files with this problem.

agcapps commented on src/axom/slic/streams/LumberjackStream.hpp:
```diff
@@ -109,6 +109,15 @@ class LumberjackStream : public LogStream
   virtual void write();
 
+  /*!
+   * \brief Confirms that abort flag(s) was set among all ranks.
```
Style suggestion, for the sake of clarity (take as you please):
```suggestion
   * \brief Confirms that abort flag(s) was set on one or more ranks.
```

agcapps reviewed (changes requested):
As I noted in slic_macros.hpp, if you call `slic::locErrorMessage()` it calls `ensureInitialized()` for you.  So you don't need to test `isInitialized()` to call `abortIfEnabled()`.  Also, as in `SLIC_ERROR_IF`, I think the call to `abortIfEnabled` should go inside the braces after `logErrorMessage()`.

Otherwise, looks good.

rhornung67 commented on src/axom/slic/core/LogStream.hpp:
```diff
@@ -120,6 +120,32 @@ class LogStream
    */
   virtual void push() {};
 
+  /*!
```
Should these methods be exposed in the public interface? Specifically, are there concerns about users calling them?

rhornung67 commented on src/axom/slic/core/Logger.cpp:
```diff
@@ -77,14 +77,75 @@ Logger::~Logger()
+  }
+
+  unsigned nstreams = static_cast<unsigned>(m_logStreams[level].size());
+  if(nstreams == 0)
```
The nstreams == 0 case is covered by the for-loop below, right? So it can be removed.

rhornung67 commented on src/axom/slic/core/Logger.hpp:
```diff
@@ -66,18 +66,36 @@ class Logger
-   * \brief Aborts and flushes on warning or error if corresponding AbortOnError
-   *  or AbortOnWarning is set to true
-   * \collective
+   * \brief Gets the abort flag for the given level.
```
white space between the brief statement and param statements would increase readability. Same for white space between param statements and return statements.

This comment applies to all methods in these files.

kennyweiss commented on src/axom/slic/interface/slic_macros.hpp:
```diff
@@ -45,13 +47,16 @@
     std::ostringstream __oss;                                     \
     __oss << msg;                                                 \
     axom::slic::logErrorMessage(__oss.str(), __FILE__, __LINE__); \
+    axom::slic::abortIfEnabled(axom::slic::message::Error);       \
```
If the functions are renamed, this might be
```cpp
axom::slic::signalAbort(slic::messsage::Error);
```

kennyweiss reviewed (commented):
Thanks @bmhan12 

A consequence of the bugfix in this PR is that, for MPI-based logstreams, slic macros that were _potentially_ collective (on error) are now _always_ collective to check if any rank has an error (when abort on error is enabled).

As mentioned in the review and during our meeting, this PR hides the problem that axom's internal implementation (and that of our users) use these macrosin a non-collective manner, e.g. in cases, such as the `InOutOctree` where the  ranks might be checking for different data-specific error conditions.
Setting `axom::slic::disableAbortOnWarning()` and using `SLIC_CHECK` instead of `SLIC_ASSERT` in our tests and examples effectively means that these cases will not be caught by our tests/examples, but they will be caught by our users.

Unfortunately, I think we should hold off on this PR, and revise our strategy for addressing #26.  
Some ideas:
* Perhaps the solution is to avoid flushing collectively during slic aborts, and add a local flush for these cases, so we can use the `MPI_Abort()` functionality ?
* Alternatively, within axom, we can ensure we're using a logger with only serial (non-MPI) LogStreams in cases where the behavior is rank-specific ?

My review also has some suggestions for improved function names that we should also discuss at some point.

kennyweiss commented on src/axom/slic/tests/slic_interface.cpp:
```diff
@@ -41,3 +42,74 @@ TEST(slic_interface, logging_level)
+    new axom::slic::GenericOutputStream(&std::cout));
+
+  // test abort on info/debug (no-ops)
+  axom::slic::setAbortFlag(true, axom::slic::message::Info);
+  EXPECT_FALSE(axom::slic::getAbortFlag(axom::slic::message::Info));
```
Note: Looks like @agcapps had similar suggestions elsewhere on this PR.

bmhan12 commented:
> Unfortunately, I think we should hold off on this PR, and revise our strategy for addressing #26. Some ideas:
> 
> * Perhaps the solution is to avoid flushing collectively during slic aborts, and add a local flush for these cases, so we can use the `MPI_Abort()` functionality ?
> * Alternatively, within axom, we can ensure we're using a logger with only serial (non-MPI) LogStreams in cases where the behavior is rank-specific ?

@kennyweiss  @white238 
I agree, in the PR's current state it's not a good solution to #26. 

I'm in favor of adding a local flush, and am working on that as a solution.

My question is if we want to have the option for both abort behaviors? (Collective flush then MPI_Abort(), Single-rank flush then MPI_Abort())

We could potentially have a toggle for the macros between these two behaviors, with the single-rank flush behavior being default. Or, we could have like `SLIC_ALL_ERROR` and `SLIC_ERROR`  for collective and local flushes respectively.

rhornung67 commented:
FWIW, I think the local flush followed by MPI_Abort() is the most robust solution and is easiest for users. The will see information from ranks that called flush and all ranks will exit gracefully. The collective flush option feels to me like the current situation we are in and requires careful coordination across all ranks by users.

bmhan12 commented:
Updates (current PR state):

- Removed collective call to check ranks for abort flags before flushing. The aborting rank now flushes locally then calls MPI_Abort().
- abortIfEnabled() simplified to abort(), which just calls system abort or MPI_Abort().
- Re-enabled abort on error for examples that were disabled before in this PR.

kennyweiss reviewed (changes requested):
Thanks for updating this @bmhan12 -- this is a huge improvement and I think we're almost there.

My main questions/concerns at this point are:
* clarity/expressing intent, i.e. whether we should add a flag to `flushSteams()` (as in the current solution) or if we should add a separate function, like `flushLocalStreams()` / `flushStreamsLocally()`
* are we flushing on the non-error streams before exiting after an aborting slic macro? It seems like we're currently not doing that, but we probably should.

rhornung67 commented on src/axom/slic/core/LogStream.hpp:
```diff
@@ -103,15 +105,21 @@ class LogStream
    *  operation intended for a synchronization checkpoint.
    */
-  virtual void flush() {};
+  virtual void flush(bool AXOM_UNUSED_PARAM(single_rank) = false) {};
```
@kennyweiss FWIW, I think localFlush() is probably the cleanest. Then, make it clear/obvious in the docs that flush is collective for MPI-based streams?

bmhan12 replied:
@rhornung67  @kennyweiss Agreed. I like `localFlush()` for LogStream and` flushLocalStreams()` for the slic interface.

bmhan12 replied:
Applied Changes.

rhornung67 commented on src/axom/slic/interface/slic.hpp:
```diff
@@ -312,18 +308,46 @@ void logWarningMessage(const std::string& message,
+ *  user-supplied communicator must call this method.
  */
-void flushStreams();
+void flushStreams(bool single_rank = false);
```
Change to 'local = false' for consistency?

bmhan12 commented on src/axom/slic/interface/slic.hpp:
```diff
@@ -188,47 +208,23 @@ void addStreamToMsgLevel(GenericOutputStream* ls, message::Level level);
 /*!
  * \brief Logs the given message to all registered streams.
- * \collective
+ *
```
I enabled ReadTheDocs for this branch [here](https://axom.readthedocs.io/en/feature-han12-slicexit/index.html#).

From the [doxygen page](https://axom.readthedocs.io/en/feature-han12-slicexit/doxygen/html/collective_list.html), logMessage() no longer appears in the collective section.

white238 commented on src/axom/slic/core/Logger.cpp:
```diff
@@ -78,15 +78,7 @@ Logger::~Logger()
-    m_abortFunction();
-  }
-}
+void Logger::abort() { m_abortFunction(); }
```
This change has problems for downstream users.  Since this was collective and is no longer collective.

I am going to have to rethink how we handle this in Serac.  Though we haven't had a problem currently with any calling of error in non-collective ways. To my knowledge we have been good about it so far.

We flush SLIC before terminating it:

```
void exitGracefully(bool error)
{
  if (axom::slic::isInitialized()) {
    serac::logger::flush();
    serac::logger::finalize();
  }

  int mpi_initialized = 0;
  MPI_Initialized(&mpi_initialized);
  int mpi_finalized = 0;
  MPI_Finalized(&mpi_finalized);
  if (mpi_initialized && !mpi_finalized) {
    MPI_Finalize();
  }
  profiling::finalize();

  accelerator::terminateDevice();

  error ? exit(EXIT_FAILURE) : exit(EXIT_SUCCESS);
}

bmhan12 commented on src/axom/slic/interface/slic_macros.hpp:
```diff
@@ -69,13 +78,20 @@
+      if(axom::slic::isAbortOnErrorsEnabled())                      \
+      {                                                             \
+        axom::slic::flushStreams(true);                             \
+        axom::slic::abort();                                        \
```
> are we flushing on the non-error streams before exiting after an aborting slic macro? It seems like we're currently not doing that, but we probably should.

@kennyweiss  Are you suggesting something like this, where we always flush for the XXX_IF and XXX_ROOT macros?

```
    if(EXP)                                                         \
    {                                                               \
      std::ostringstream __oss;                                     \
      __oss << msg;                                                 \
      axom::slic::logErrorMessage(__oss.str(), __FILE__, __LINE__); \
      if(axom::slic::isAbortOnErrorsEnabled())                      \
      {                                                             \
        axom::slic::flushStreams(true);                             \
        axom::slic::abort();                                        \
      }                                                             \
    }                                                               \
    // Added - Always flush
    else{                                                           \
        axom::slic::flushStreams(true);                             \
    }                                                               \
  } while(axom::slic::detail::false_value)
```

bmhan12 replied:
Moved to abort handler.

bmhan12 commented on src/axom/slic/tests/slic_lumberjack_parallel.cpp:
```diff
@@ -0,0 +1,359 @@
+
+  }  // end for loop
+
+  slic::disableAbortOnWarning(); /* disable abort for testing purposes */
```
Added.

kennyweiss commented on src/axom/slic/core/LogStream.hpp:
```diff
@@ -102,7 +104,17 @@ class LogStream
-   * \brief Flushes the log stream. It's a NO-OP by default.
+   * \brief Flushes the log stream on current rank. It's a NO-OP by default.
+   *
+   * \note The intent of this method is to be overridden by concrete
+   *  implementations. This is primarily useful for applications running
+   *  in a distributed MPI environment.
```
I'm not sure if this `note` needs to be there.

On the other hand, it might be good to explicitly call out the fact that this function is not collective.

kennyweiss commented on src/axom/slic/core/Logger.hpp:
```diff
@@ -257,7 +246,38 @@ class Logger
 
   /*!
-   * \brief Flushes all streams.
+   * \brief Flushes all streams for current rank.
```
Here too, it might be good to call out that this call is not collective, and should generally be avoided in favor of  `flushStreams` unless a rank is aborting since it could potentially not play nicely with all `LogStream`s.

kennyweiss commented on src/axom/slic/streams/SynchronizedStream.hpp:
```diff
@@ -73,7 +73,15 @@ class SynchronizedStream : public LogStream
 
   /*!
-   * \brief Dumps the messages to the console in rank-order.
+   * \brief Dumps the messages to the console in rank-order for
+   *        just the current rank
```
I think we lose `rank-order` here.

Consider changing this to something like:
```suggestion
   * \brief Dumps the current rank's messages to the console (non-collectively)
```

kennyweiss reviewed (approved):
Thanks for working through all the suggestions/issues @bmhan12 !

white238 commented on src/axom/slic/core/LogStream.hpp:
```diff
@@ -102,7 +104,20 @@ class LogStream
 
   /*!
-   * \brief Flushes the log stream. It's a NO-OP by default.
+   * \brief Flushes the log stream on current rank. It's a NO-OP by default.
```
Is this clear enough that it is flushing localing to the output stream and not down the SLIC stream? I tried to word it a couple times but it wasn't helping.  I could spend more time on it if you like.

bmhan12 replied:
> Is this clear enough that it is flushing localing to the output stream and not down the SLIC stream? I tried to word it a couple times but it wasn't helping. I could spend more time on it if you like.

I gave it another pass, and used the term "console" to try and make it more clear that it is going directly to the output stream.

white238 commented on src/axom/slic/core/Logger.cpp:
```diff
@@ -237,8 +233,21 @@ void Logger::logMessage(message::Level level,
+}
+
+//------------------------------------------------------------------------------
+void Logger::flushLocalStreams()
```
localFlushStreams? to match `localFlush`

white238 replied:
Yet another name idea:

outputLocalMessages

white238 commented on src/axom/slic/interface/slic_macros.hpp:
```diff
@@ -129,10 +145,10 @@
-      else                                                            \
-      {                                                               \
-        axom::slic::abortIfEnabled(axom::slic::message::Error);       \
+        if(axom::slic::isAbortOnErrorsEnabled())                      \
```
This needs to about if not on root as well

bmhan12 replied:
> This needs to abort if not on root as well

Yes, but I don't think that can be done until we have a better collective/non-collective solution (unless I'm missing the obvious solution here...).

Non-root ranks would have to be notified that a root rank is aborting.

white238 commented on src/axom/slic/tests/slic_macros_parallel.cpp:
```diff
@@ -569,7 +861,9 @@ int main(int argc, char* argv[])
 
   slic::addStreamToAllMsgLevels(
-    new slic::GenericOutputStream(&slic::internal::test_stream, msgfmt));
+    new slic::SynchronizedStream(&slic::internal::test_stream,
```
Good catch~

white238 commented on src/axom/slic/tests/slic_macros_parallel.cpp:
```diff
@@ -454,97 +500,343 @@ TEST(slic_macros_parallel, test_debug_macros)
-  check_line(slic::internal::test_stream.str(), (__LINE__ - 7));
-  slic::internal::clear();
+  check_line(slic::internal::test_stream.str(), (__LINE__ - 8));
+  slic::internal::clear_stream();
```
Should this be plural, `clear_streams`?

white238 commented on src/axom/slic/tests/slic_lumberjack_parallel.cpp:
```diff
@@ -0,0 +1,395 @@
+// UNIT TESTS
+//------------------------------------------------------------------------------
+//------------------------------------------------------------------------------
+TEST(slic_lumberjack_parallel, test_abort_error_macros)
```
If this is just a copy of the parallel tests, we could move this to a templated testing suite...

bmhan12 replied:
> If this is just a copy of the parallel tests, we could move this to a templated testing suite...

It is, the main difference being these parallel lumberjack tests are only defined if axom is built with lumberjack. 
If I combine the two, the parallel tests will be defined only if axom is built with slic and lumberjack. 
I think it'll make future maintenance easier to combine them, just noting this here so I remember later.

bmhan12 commented:
Updates:

- Renamed: `localFlush()` --> `outputLocal()`
- Renamed: `flushLocalStreams()` --> `outputLocalMessages()`
- Templated slic parallel tests for `SynchronizedStream `and `LumberjackStream`
- Added a warning in sphinx about aborting and `outputLocalMessages()` : https://axom.readthedocs.io/en/feature-han12-slicexit/axom/slic/docs/sphinx/sections/architecture.html

white238 commented:
@bmhan12 can you do one more thing? Can you add a comment to `setAbortFunction()` that no collective calls should be made in this function?

Then I need to verify a case that will cause problems for Serac and this should be good to go!  Thanks for holding on through the long review/discussion of what we really wanted!

bmhan12 commented:
>  Can you add a comment to setAbortFunction() ...

Added.

No problem!

white238 commented on src/axom/slic/interface/slic.hpp:
```diff
@@ -313,17 +312,52 @@ void logWarningMessage(const std::string& message,
+//! All ranks in the user-supplied communicator must call the method
+//! when used within an MPI distributed environment.
+//!
+//! \sa axom::slic::isAbortOnErrorsEnabled()
```
Is calling these collective correct? I know they have to be called on each node to be correct but there is no communication, its just getting/setting a local boolean.

bmhan12 replied:
Good catch.

white238 commented:
I ran this through a toy test by hand in an attempt to make sure it works how we think it works.  It did!

It also reminded me I didn't set a state variable for the root macros in Serac!

Thanks @bmhan12 ! I appreciate the effort you put into this. I know it was long and winding but the end result is very good as well as the lessons learned.

rhornung67 reviewed (approved):
Thanks @bmhan12 Nice work!
